File: stacks/find-the-most-competitive-subsequence/find-the-most-competitive-subsequence.java
class Solution {
  public int[] mostCompetitive(int[] nums, int k) {
    Stack<Integer> stack = new Stack<>();
    for (int i = 0; i < nums.length; i++) {
      while (!stack.empty() && stack.peek() > nums[i] && k < stack.size() + nums.length - i) {
        stack.pop();
      }
      if (stack.size() < k) {
        stack.push(nums[i]);
      }
    }

    int[] arr = new int[stack.size()];

    for (int i = k - 1; i >= 0; i--) {
      arr[i] = stack.pop();
    }

    return arr;
  }
}

File: stacks/find-the-most-competitive-subsequence/find-the-most-competitive-subsequence.ts
function mostCompetitive(nums: number[], k: number): number[] {
  const stack: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    while (
      stack.length &&
      stack[stack.length - 1] > nums[i] &&
      k < stack.length + nums.length - i
    ) {
      stack.pop();
    }
    if (stack.length < k) {
      stack.push(nums[i]);
    }
  }

  const arr: number[] = new Array(k);

  for (let i = k - 1; i >= 0; i--) {
    arr[i] = stack.pop()!;
  }

  return arr;
}
